infer/src/integration/CaptureSILJson.ml

```diff
@@ -0,0 +1,17 @@
+(*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *)
+
+(** Capture module for the json analysis in the capture phase *)
+
+open! IStd
+
+let capture ~changed_files ~cfg_json ~tenv_json =
+  let tenv = InferAnalyzeJson.parse_tenv (Yojson.Safe.from_file tenv_json) in
```

Comment:
Maybe this is a good point to set `Language.curr_language` to `CIL` instead of during the analysis, if it's needed at all (does anything break without?). The clang and Java frontends do the same, though I confess I'm not sure why exactly they need to (eg the Erlang frontend doesn't set the current language).

Reply:
Yes, if we don't set `Language.curr_language` to `CIL`, some of our CIL-specific logics will be skipped and ended up with some bug reports missing. 

I found that if we run `infer run` command, set `Language.curr_language` to `CIL` in `CaptureSILJson` will work just fine for us. However, when we split `infer run` into `infer capture` and `infer analyze`, set `Language.curr_language` to `CIL` in  `CaptureSILJson` won't be fine for us because the language set in `infer capture` won't be passed to `infer analyze`. That is the reason why we ended up creating a `--cli-lang` flag. Do you have any better suggestion for this?

Reply:
I've done some digging: the issue seems to be in the way that we disable filtering for C# in JsonReports.ml:
https://github.com/facebook/infer/blob/b67e2d518dd67576e248785e72cb9ebd1a508607/infer/src/integration/JsonReports.ml#L75

This relies on the global `curr_language` being set to `CIL` but we are outside any particular analysis here (this is run after all analyses have run and collects all the summaries of all the procedures; conceivably there could be several languages represented here so `Language.curr_language` doesn't make sense any more). We should test the language of the current proc name for which we want to report the issue instead:

```diff
diff --git a/infer/src/integration/JsonReports.ml b/infer/src/integration/JsonReports.ml
index 196c857ea3..9691b211d6 100644
--- a/infer/src/integration/JsonReports.ml
+++ b/infer/src/integration/JsonReports.ml
@@ -71,8 +71,8 @@ let loc_trace_to_jsonbug_record trace_list ekind =
       record_list
 
 
-let should_report issue_type error_desc =
-  if (not Config.filtering) || Language.curr_language_is CIL then true
+let should_report proc_name issue_type error_desc =
+  if (not Config.filtering) || Language.equal (Procname.get_language proc_name) CIL then true
   else
     let issue_type_is_null_deref =
       let null_deref_issue_types =
@@ -187,7 +187,7 @@ module JsonIssuePrinter = MakeJsonListPrinter (struct
     if
       error_filter source_file err_key.issue_type
       && should_report_proc_name
-      && should_report err_key.issue_type err_key.err_desc
+      && should_report proc_name err_key.issue_type err_key.err_desc
       && not (is_in_clang_header source_file)
     then
       let severity = IssueType.string_of_severity err_key.severity in
```

This filtering is only about one special case: null deref issues reported by biabduction, and CIL issues get filtered out because they have no "bucket" set for them. Buckets are a concept in the biabduction analysis used to represent the "confidence" we have in the issue to report, and if no bucket is set then by default we don't report the issue:
https://github.com/facebook/infer/blob/main/infer/src/integration/JsonReports.ml#L84
https://github.com/facebook/infer/blob/main/infer/src/absint/Localise.ml#L86-L89

Could you try the fix above and remove the config option? The tests pass for me if I do this. I'd rather not have the option if we don't need it, and also because infer should work even if we mix several languages in the same analysis and this breaks that (not a deal breaker since I doubt this is very useful at the moment but still, no reason to break it unless we need to).

Reply:
Thanks for your comments and advice! I have adopted your changes and removed config, which passed our tests as well.